[PyTorch] grouped linear with fp8 current scaling fails with exception: Unable to find suitable cuBLAS grouped GEMM algorithm

**Describe the bug**

https://github.com/NVIDIA/TransformerEngine/commit/25ad5cb904c602d04291d88ee04c6f6f3ea3467d

When running the grouped linear test with CUDA graph safe mode and FP8 current scaling (test_grouped_linear_fused_path_cuda_graph_safe[True-fp8_current_scaling]) on H100 (SM 90), the test fails with a cuBLAS assertion error indicating that no suitable grouped GEMM algorithm can be found.



**Steps/Code to reproduce bug**

I noticed that recent commits have added FP8 current scaling with graph capture support for compute capability 9.0, and I'm trying to use this feature on H100. However, the test fails on the grouped GEMM algorithm selection.

1. Docker Environment:

``` bash
docker pull nvcr.io/nvidia/pytorch:26.04-py3
docker run --gpus all -it nvcr.io/nvidia/pytorch:26.04-py3
```

2. Transformer Engine Setup:

```
# Clone and checkout the commit
git clone https://github.com/NVIDIA/TransformerEngine.git
cd TransformerEngine
git checkout 7cb8b313d55021eb12c5efd4d02f4c8ff7679453

# Revert NCCL-related commits (due to compatibility issues with PyTorch 26.04 image)
git revert 90baf028905d23e133ecb5920a6a68c82f1ceeb7
git revert 4955320121e500db98f1c08d8d90075c17d9469e
git revert 3df5e19b1f24325b3bfd34bca10717aaba86b960
git revert 4cd244eee50b2ac3b12ed2f77ac02964ca99f6c5

# Build and install
NVTE_BUILD_THREADS_PER_JOB=8 NVTE_FRAMEWORK=pytorch python setup.py bdist_wheel  -vvv
```

3. Run the failing test:

``` bash
pytest -svv tests/pytorch/test_grouped_linear.py -k 'test_grouped_linear_fused_path_cuda_graph_safe'
```

log:
```
pytest -svv tests/pytorch/test_grouped_linear.py -k 'test_grouped_linear_fused_path_cuda_graph_safe'
===================================== test session starts =====================================
platform linux -- Python 3.12.3, pytest-9.1.1, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase(PosixPath('/SharedData/luw/workspace/dsv4/TransformerEngine/.hypothesis/examples'))
rootdir: /SharedData/luw/workspace/dsv4/TransformerEngine
configfile: pyproject.toml
plugins: darkgraylib-2.4.1, random-order-1.2.0, mock-3.15.1, asyncio-1.4.0, cov-7.1.0, typeguard-4.5.1, anyio-4.13.0, hypothesis-6.130.8
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 1411 items / 1403 deselected / 8 selected

tests/pytorch/test_grouped_linear.py::test_grouped_linear_fused_path_cuda_graph_safe[True-bf16] PASSED
tests/pytorch/test_grouped_linear.py::test_grouped_linear_fused_path_cuda_graph_safe[True-fp8_current_scaling] FAILED
tests/pytorch/test_grouped_linear.py::test_grouped_linear_fused_path_cuda_graph_safe[True-mxfp8] SKIPPED
tests/pytorch/test_grouped_linear.py::test_grouped_linear_fused_path_cuda_graph_safe[True-nvfp4] SKIPPED
tests/pytorch/test_grouped_linear.py::test_grouped_linear_fused_path_cuda_graph_safe[False-bf16] PASSED
tests/pytorch/test_grouped_linear.py::test_grouped_linear_fused_path_cuda_graph_safe[False-fp8_current_scaling] PASSED
tests/pytorch/test_grouped_linear.py::test_grouped_linear_fused_path_cuda_graph_safe[False-mxfp8] SKIPPED
tests/pytorch/test_grouped_linear.py::test_grouped_linear_fused_path_cuda_graph_safe[False-nvfp4] SKIPPED

========================================== FAILURES ===========================================
__________ test_grouped_linear_fused_path_cuda_graph_safe[True-fp8_current_scaling] ___________

fp8_recipe = recipe_type=Float8CurrentScaling, format=HYBRID, fp8_quant_fwd_inp=Qparams(
power_2_scale=False,
amax_epsilon=0.0,
ran...ulator=True), fp8_gemm_wgrad=MMParams(use_split_accumulator=True), fp8_dpa=False, fp8_mha=False, backward_override=None
bias = True, monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7f181cb42360>

    @pytest.mark.parametrize(
        "fp8_recipe",
        [
            None,
            pytest.param(
                recipe.Float8CurrentScaling(),
                marks=pytest.mark.skipif(not _fp8_available, reason=_reason_for_no_fp8),
            ),
            pytest.param(
                recipe.MXFP8BlockScaling(),
                marks=pytest.mark.skipif(not _mxfp8_available, reason=_reason_for_no_mxfp8),
            ),
            pytest.param(
                recipe.NVFP4BlockScaling(disable_stochastic_rounding=True),
                marks=pytest.mark.skipif(not _nvfp4_available, reason=_reason_for_no_nvfp4),
            ),
        ],
        ids=["bf16", "fp8_current_scaling", "mxfp8", "nvfp4"],
    )
    @pytest.mark.parametrize("bias", _ALL_BOOLEAN)
    def test_grouped_linear_fused_path_cuda_graph_safe(fp8_recipe, bias, monkeypatch):
        """Fused GroupedTensor GEMM path should be CUDA graph capturable."""
        use_fp8 = fp8_recipe is not None
        device_capability = torch.cuda.get_device_capability()
        if not (9, 0) <= device_capability <= (11, 0):
            pytest.skip(
                "GroupedTensor grouped GEMM path requires Hopper (SM90) or Blackwell (SM10x and SM110)."
            )
        # MXFP8/NVFP4 grouped quantization kernels require Blackwell, but FP8 per-tensor
        # current scaling also runs on the Hopper grouped GEMM path.
        is_current_scaling = use_fp8 and fp8_recipe.float8_current_scaling()
        if use_fp8 and not is_current_scaling and device_capability < (10, 0):
            pytest.skip(
                "Quantized GroupedTensor grouped GEMM path (MXFP8/NVFP4) requires Blackwell (SM100+)."
            )
        cublaslt_version = tex.get_cublasLt_version()
        if device_capability < (10, 0) and cublaslt_version < 130400:
            pytest.skip("Grouped GEMM on Hopper requires cuBLAS 13.4+.")
        if cublaslt_version < 130300:
            pytest.skip("Grouped GEMM requires cuBLAS 13.3+.")

        monkeypatch.setenv(_FUSED_GROUPED_GEMM_ENV, "1")
        FP8GlobalStateManager.reset()

        dtype = torch.bfloat16
        device = "cuda"
        num_gemms = 3
        in_features = 128
        out_features = 128
        split_sizes = [128, 256, 384]
        total_tokens = sum(split_sizes)
        static_m_splits = torch.tensor(split_sizes, dtype=torch.int64, device=device)

        grouped_linear = GroupedLinear(
            num_gemms,
            in_features,
            out_features,
            bias=bias,
            params_dtype=dtype,
            device=device,
        )

        static_x = torch.randn(total_tokens, in_features, dtype=dtype, device=device)
        static_x.requires_grad_(True)
        static_dy = torch.randn(total_tokens, out_features, dtype=dtype, device=device)
        static_out_buf = torch.empty(total_tokens, out_features, dtype=dtype, device=device)

        def _zero_grads():
            if static_x.grad is not None:
                static_x.grad.zero_()
            for param in grouped_linear.parameters():
                if param.grad is None:
                    param.grad = torch.zeros_like(param)
                else:
                    param.grad.zero_()

        def _clone_param_grads():
            return [param.grad.detach().clone() for param in grouped_linear.parameters()]

        def _train_step(x, dy, out_buf, *, use_graphed):
            with autocast(enabled=use_fp8, recipe=fp8_recipe):
                out = (
                    graphed_grouped_linear(x, static_m_splits)
                    if use_graphed
                    else grouped_linear(x, static_m_splits)
                )
            out.backward(dy)
            out_buf.copy_(out)
            return out_buf

>       graphed_grouped_linear = te.make_graphed_callables(
            grouped_linear,
            (static_x, static_m_splits),
            num_warmup_iters=3,
            enabled=use_fp8,
            recipe=fp8_recipe,
        )

tests/pytorch/test_grouped_linear.py:1885:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/lib/python3.12/dist-packages/transformer_engine/pytorch/graph.py:1369: in make_graphed_callables
    graphed_callables = _make_graphed_callables(
/usr/local/lib/python3.12/dist-packages/transformer_engine/pytorch/graph.py:490: in _make_graphed_callables
    outputs, _ = _tree_flatten(func(*args, **kwargs))
                               ^^^^^^^^^^^^^^^^^^^^^
/usr/local/lib/python3.12/dist-packages/transformer_engine/pytorch/graph.py:1342: in call_func
    outputs = old_call_funcs[block_cls](self, *args, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py:1778: in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py:1884: in _call_impl
    return inner()
           ^^^^^^^
/usr/local/lib/python3.12/dist-packages/torch/nn/modules/module.py:1832: in inner
    result = forward_call(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/local/lib/python3.12/dist-packages/transformer_engine/pytorch/jit.py:67: in wrapper
    return disabled_f(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/local/lib/python3.12/dist-packages/torch/_dynamo/eval_frame.py:1256: in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
/usr/local/lib/python3.12/dist-packages/transformer_engine/pytorch/module/grouped_linear.py:1796: in forward
    out, new_workspaces = linear_fn(
/usr/local/lib/python3.12/dist-packages/torch/autograd/function.py:596: in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/local/lib/python3.12/dist-packages/transformer_engine/pytorch/module/grouped_linear.py:515: in forward
    return _GroupedLinear._forward_grouped_tensor(
/usr/local/lib/python3.12/dist-packages/transformer_engine/pytorch/module/grouped_linear.py:324: in _forward_grouped_tensor
    general_grouped_gemm_for_grouped_tensor(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

A = [Float8TensorStorage(fp8_dtype=7, scale_inv=0.0002179826988140121, data=tensor([[ 0.0079, -0.0227, -0.0315,  ..., -0.0...
        [-0.0053,  0.0491,  0.0025,  ..., -0.0181, -0.0214, -0.0115]],
       device='cuda:0', dtype=torch.bfloat16))]
B = GroupedTensorStorage(num_tensors=3, shapes=None, logical_shape=[768, 128], quantizer=Float8CurrentScalingQuantizer(rowwise_usage=True, columnwise_usage=True, internal=True, ), dtype=torch.bfloat16)
out = GroupedTensorStorage(num_tensors=3, shapes=None, logical_shape=(768, 128), quantizer=None, dtype=torch.bfloat16)
layout = 'TN', accumulate = False, use_split_accumulator = False
bias = GroupedTensorStorage(num_tensors=3, shapes=[(1, 128), (1, 128), (1, 128)], logical_shape=(3, 128), quantizer=None, dtype=torch.bfloat16)
bias_scale = None, grad = False, alpha = tensor([1.], device='cuda:0')
beta = tensor([0.], device='cuda:0')

    def general_grouped_gemm_for_grouped_tensor(
        A,
        B,
        out,
        *,
        layout: str = "TN",
        accumulate: bool = False,
        use_split_accumulator: bool = False,
        bias=None,
        bias_scale: Optional[torch.Tensor] = None,
        grad: bool = False,
        alpha: Optional[torch.Tensor] = None,
        beta: Optional[torch.Tensor] = None,
    ) -> Union[torch.Tensor, List[torch.Tensor]]:
        """
        Grouped GEMM using GroupedTensor inputs.

        This uses nvte_grouped_gemm and supports different per-matrix shapes.

        The caller must ensure that GroupedTensor metadata is already compatible with the
        underlying GEMM implementation (e.g., aligned offsets and output metadata layout).
        """
        assert layout in ("TN", "NN", "NT"), f"GEMM layout {layout} not supported."
        if grad:
            raise NotImplementedError("grad is not supported for grouped_tensor GEMM yet.")
        transa = layout[0] == "T"
        transb = layout[1] == "T"
        is_discrete_out = isinstance(out, list)
        is_discrete_in = isinstance(A, list)
        if is_discrete_in and is_discrete_out:
            raise ValueError("Both A and out are discrete. This is not supported yet.")

        if isinstance(A, GroupedTensorStorage) and A.row_scaled_nvfp4:
            raise NotImplementedError("Row-scaled NVFP4 GroupedTensor GEMM is not supported yet.")
        if isinstance(B, GroupedTensorStorage) and B.row_scaled_nvfp4:
            raise NotImplementedError("Row-scaled NVFP4 GroupedTensor GEMM is not supported yet.")
        if isinstance(out, GroupedTensorStorage) and out.row_scaled_nvfp4:
            raise NotImplementedError("Row-scaled NVFP4 GroupedTensor GEMM is not supported yet.")

        if is_discrete_out:
            # wgrad case.
            grouped_gemm_impl = tex.te_general_grouped_gemm_for_discrete_out
        elif is_discrete_in:
            # Use-case: forward pass with list of weights.
            grouped_gemm_impl = tex.te_general_grouped_gemm_for_discrete_in
        else:
            # Use-case: Single Grouped Parameter for Weight/ Weight Grads.
            grouped_gemm_impl = tex.te_general_grouped_gemm_for_grouped_tensor

        if is_discrete_out and bias is not None:
            raise ValueError(
                "Bias is not supported when out is a list (discrete_out mode) yet. "
                "Apply bias manually after the GEMM."
            )

        if bias_scale is not None and bias is None:
            raise ValueError("bias_scale requires bias to be provided.")

        num_tensors = B.num_tensors
        rowwise = B.rowwise_data
        device = rowwise.device if rowwise is not None else B.columnwise_data.device

        # Hopper (SM90) uses a single shared alpha/beta scalar;
        # Blackwell+ (SM100) supports per-group alpha/beta arrays.
        per_group = torch.cuda.get_device_capability() >= (10, 0)
        num_alphabeta = num_tensors if per_group else 1

        if alpha is None:
            alpha = _get_fp32_ones_tensor(num_alphabeta, device)
        if beta is None:
            if accumulate:
                beta = _get_fp32_ones_tensor(num_alphabeta, device)
            else:
                beta = _get_fp32_zeros_tensor(num_alphabeta, device)

        if not alpha.is_cuda or not beta.is_cuda:
            raise ValueError("alpha and beta must be CUDA tensors.")

        workspace_setup = torch.empty(
            get_grouped_gemm_setup_workspace_size(num_tensors),
            dtype=torch.uint8,
            device=device,
        )
        workspace_cublas = torch.empty(
            get_cublas_workspace_size_bytes(),
            dtype=torch.uint8,
            device=device,
        )

        sm_count = get_sm_count()
        sm_count = sm_count - int(os.getenv("NVTE_EXT_MARGIN_SM", str(sm_count)))

>       return grouped_gemm_impl(
            A,
            transa,
            B,
            transb,
            out,
            bias,
            bias_scale,
            alpha,
            beta,
            workspace_setup,
            workspace_cublas,
            use_split_accumulator,
            sm_count,
        )
E       RuntimeError: /SharedData/luw/workspace/dsv4/TransformerEngine/transformer_engine/common/gemm/cublaslt_grouped_gemm.cu:995 in function select_grouped_gemm_algo: Assertion failed: status != CUBLAS_STATUS_NOT_SUPPORTED. Unable to find suitable cuBLAS grouped GEMM algorithm

/usr/local/lib/python3.12/dist-packages/transformer_engine/pytorch/cpp_extensions/gemm.py:530: RuntimeError
====================================== warnings summary =======================================
../../../../../usr/local/lib/python3.12/dist-packages/torch/cuda/__init__.py:61
  /usr/local/lib/python3.12/dist-packages/torch/cuda/__init__.py:61: FutureWarning: The pynvml package is deprecated. Please install nvidia-ml-py instead. If you did not install pynvml directly, please report this to the maintainers of the package that installed pynvml for you.
    import pynvml  # type: ignore[import]

../../../../../usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:1488
../../../../../usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:1488
  /usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:1488: DeprecationWarning: `torch.jit.script` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

tests/pytorch/test_grouped_linear.py::test_grouped_linear_fused_path_cuda_graph_safe[True-bf16]
  /usr/local/lib/python3.12/dist-packages/torch/autograd/graph.py:877: UserWarning: The AccumulateGrad node's stream does not match the stream of the node that produced the incoming gradient. This may incur unnecessary synchronization and break CUDA graph capture if the AccumulateGrad node's stream is the default stream. This mismatch is caused by an AccumulateGrad node created prior to the current iteration being kept alive. This can happen if the autograd graph is still being kept alive by tensors such as the loss, or if you are using DDP, which will stash a reference to the node. To resolve the mismatch, delete all references to the autograd graph or ensure that DDP initialization is performed under the same stream as subsequent forwards. If the mismatch is intentional, you can use torch.autograd.graph.set_warn_on_accumulate_grad_stream_mismatch(False) to suppress this warning. (Triggered internally at /opt/pytorch/pytorch/torch/csrc/autograd/input_buffer.cpp:240.)
    return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================== short test summary info ===================================
FAILED tests/pytorch/test_grouped_linear.py::test_grouped_linear_fused_path_cuda_graph_safe[True-fp8_current_scaling] - RuntimeError: /SharedData/luw/workspace/dsv4/TransformerEngine/transformer_engine/common/gemm/cublaslt_grouped_gemm.cu:995 in function select_grouped_gemm_algo: Assertion failed: status != CUBLAS_STATUS_NOT_SUPPORTED. Unable to find suitable cuBLAS grouped GEMM algorithm
============= 1 failed, 3 passed, 4 skipped, 1403 deselected, 4 warnings in 4.41s =============
```


**Expected behavior**

The test should pass on H100 (compute capability 9.0) when using FP8 current scaling with CUDA graph safe mode, as recent commits claim to support this feature for SM 90 devices.

**Environment overview (please complete the following information)**

Environment location: Docker ([nvcr.io/nvidia/pytorch:26.04-py3](https://nvcr.io/nvidia/pytorch:26.04-py3))

Method of Transformer Engine install: From source (commit 7cb8b313d55021eb12c5efd4d02f4c8ff7679453)

Docker pull & run commands:

``` bash
docker pull nvcr.io/nvidia/pytorch:26.04-py3
docker run --gpus all -it nvcr.io/nvidia/pytorch:26.04-py3
```

**Environment details**

**Device details**
- NVIDIA H100 80GB HBM3

**Additional context**

NCCL-related commits reverted:
90baf028905d23e133ecb5920a6a68c82f1ceeb7

4955320121e500db98f1c08d8d90075c17d9469e

3df5e19b1f24325b3bfd34bca10717aaba86b960

4cd244eee50b2ac3b12ed2f77ac02964ca99f6c5

These commits were reverted because they depend on NCCL v2.30+, which is not compatible with the PyTorch version in the current Docker image.

Questions:
Is the grouped GEMM algorithm selection for FP8 current scaling dependent on NCCL version or the NCCL-related commits I reverted?

Could the revert of these commits have broken the FP8 grouped GEMM path indirectly, even though they are NCCL-related and not directly related to cuBLAS?

Is there a known cuBLAS version requirement for FP8 current scaling grouped GEMM on H100? The container's CUDA version is 13.2.78.

Could there be an issue with the graph capture metadata when using FP8 current scaling that causes cuBLAS to reject the algorithm?

Any guidance on how to resolve this issue or additional debugging steps would be greatly appreciated!

--- a/tests/cpp/operator/test_grouped_gemm.cu
+++ b/tests/cpp/operator/test_grouped_gemm.cu
@@ -234,6 +234,8 @@ std::vector<std::tuple<size_t, size_t, size_t>> make_shapes(ShapeCase scase) {
 
 constexpr size_t kCublasGroupedGemmVersion = 130300;        // Blackwell-only grouped GEMM
 constexpr size_t kCublasGroupedGemmHopperVersion = 130400;  // adds Hopper support
+// FP8 per-tensor scaling grouped GEMM on Hopper requires cuBLAS 13.5+
+constexpr size_t kCublasFp8TensorScalingGroupedGemmHopperVersion = 130500;
 constexpr size_t kCublasWorkspaceBytes = 32ull * 1024 * 1024;
 
 inline std::string grouped_gemm_skip_reason(const TestParams& params) {
@@ -256,8 +258,15 @@ inline std::string grouped_gemm_skip_reason(const TestParams& params) {
     const bool is_blackwell_plus = cc >= blackwellComputeCapability;
     const bool fp8_block = is_fp8_block_recipe(params.recipe);
     if (!is_blackwell_plus && !fp8_block) {
-      return std::string(recipe_name(params.recipe)) +
-             " grouped GEMM requires Blackwell (SM100) or newer, " + cc_suffix;
+      if (params.recipe == InputRecipe::kFP8Current) {
+        if (cublas_ver < kCublasFp8TensorScalingGroupedGemmHopperVersion) {
+          return "FP8 per-tensor scaling grouped GEMM on Hopper (SM90) requires cuBLAS 13.5+, "
+                 "but run-time cuBLAS version is " + std::to_string(cublas_ver) + ".";
+        }
+      } else {
+        return std::string(recipe_name(params.recipe)) +
+               " grouped GEMM requires Blackwell (SM100) or newer, " + cc_suffix;
+      }
     }
     if (is_blackwell_plus && fp8_block) {
       return "FP8 block scaling grouped GEMM is only supported on Hopper (SM90), " + cc_suffix;

--- a/tests/pytorch/test_grouped_linear.py
+++ b/tests/pytorch/test_grouped_linear.py
@@ -1615,6 +1615,8 @@ def test_grouped_linear_grouped_tensor_path_matches_legacy(
     cublaslt_version = tex.get_cublasLt_version()
     if device_capability < (10, 0) and cublaslt_version < 130400:
         pytest.skip("Grouped GEMM on Hopper requires cuBLAS 13.4+.")
+    if is_current_scaling and device_capability < (10, 0) and cublaslt_version < 130500:
+        pytest.skip("FP8 per-tensor scaling grouped GEMM on Hopper requires cuBLAS 13.5+.")
     if cublaslt_version < 130300:
         pytest.skip("Grouped GEMM requires cuBLAS 13.3+.")
 
@@ -1830,6 +1832,8 @@ def test_grouped_linear_fused_path_cuda_graph_safe(fp8_recipe, bias, monkeypatch
     cublaslt_version = tex.get_cublasLt_version()
     if device_capability < (10, 0) and cublaslt_version < 130400:
         pytest.skip("Grouped GEMM on Hopper requires cuBLAS 13.4+.")
+    if is_current_scaling and device_capability < (10, 0) and cublaslt_version < 130500:
+        pytest.skip("FP8 per-tensor scaling grouped GEMM on Hopper requires cuBLAS 13.5+.")
     if cublaslt_version < 130300:
         pytest.skip("Grouped GEMM requires cuBLAS 13.3+.")
 

--- a/transformer_engine/common/gemm/cublaslt_grouped_gemm.cu
+++ b/transformer_engine/common/gemm/cublaslt_grouped_gemm.cu
@@ -43,6 +43,10 @@ inline void CreateCublasHandle(cublasLtHandle_t *handle) {
 // FP8 block scaling support for grouped GEMM requires cuBLAS 13.4+
 #define CUBLAS_FP8_BLOCK_GROUPED_GEMM_VERSION 130400
 
+// FP8 tensor scaling (per-tensor current/delayed scaling) support for grouped GEMM
+// on Hopper (SM90) requires cuBLAS 13.5+
+#define CUBLAS_FP8_TENSOR_SCALING_GROUPED_GEMM_HOPPER_VERSION 130500
+
 // BF16 support for grouped GEMM requires cuBLAS 13.3+
 #define CUBLAS_GROUPED_GEMM_VERSION 130300
 
@@ -1054,6 +1058,14 @@ inline void execute_grouped_gemm(const GroupedGemmSetupWorkspace &setup_workspac
                                          setup_workspace.b_scale_inv_ptrs, A_sel.scaling_mode,
                                          B_sel.scaling_mode);
   } else if (config.use_fp8) {
+    const int sm = transformer_engine::cuda::sm_arch(transformer_engine::cuda::current_device());
+    if (sm < 100) {
+      NVTE_CHECK(transformer_engine::cuda::cublas_version() >=
+                     CUBLAS_FP8_TENSOR_SCALING_GROUPED_GEMM_HOPPER_VERSION,
+                 "FP8 tensor scaling grouped GEMM on Hopper (SM90) requires cuBLAS 13.5+, "
+                 "but run-time cuBLAS version is ",
+                 transformer_engine::cuda::cublas_version());
+    }
     set_fp8_scale_pointers(matmulDesc, setup_workspace.a_scale_inv_ptrs,
                            setup_workspace.b_scale_inv_ptrs);
   }

--- a/transformer_engine/pytorch/module/grouped_linear.py
+++ b/transformer_engine/pytorch/module/grouped_linear.py
@@ -108,6 +108,8 @@ class _GroupedLinear(torch.autograd.Function):
           per-tensor current scaling.
         FP8 delayed scaling and FP8 block scaling are not supported because the
         corresponding grouped quantization kernels are missing.
+        Grouped GEMM requires cuBLAS 13.3+ (13.4+ on Hopper, 13.5+ for FP8
+        per-tensor current scaling on Hopper); otherwise the legacy path is used.
         Non-RHT NVFP4 falls back to the legacy path because graph-safe grouped quantization
         currently requires RHT.
 
@@ -126,8 +128,14 @@ class _GroupedLinear(torch.autograd.Function):
             or save_original_input
         ):
             return False
-        # 3. Filter by compute capability
-        if not (9, 0) <= get_device_compute_capability() <= (11, 0):
+        # 3. Filter by compute capability and cuBLAS version
+        device_capability = get_device_compute_capability()
+        if not (9, 0) <= device_capability <= (11, 0):
+            return False
+        cublaslt_version = tex.get_cublasLt_version()
+        if cublaslt_version < 130300:
+            return False
+        if device_capability < (10, 0) and cublaslt_version < 130400:
             return False
         # 4. Output quantization is not supported.
         if any(q is not None for q in output_quantizers):
@@ -135,9 +143,12 @@ class _GroupedLinear(torch.autograd.Function):
         # 5. Filter by quantization recipes.
         if fp8:
             if all(isinstance(q, Float8CurrentScalingQuantizer) for q in input_quantizers):
+                # FP8 per-tensor scaling grouped GEMM on Hopper requires cuBLAS 13.5+.
+                if device_capability < (10, 0) and cublaslt_version < 130500:
+                    return False
                 return True
             # MXFP8 and NVFP4 require Blackwell+.
-            if not (10, 0) <= get_device_compute_capability() <= (11, 0):
+            if not (10, 0) <= device_capability <= (11, 0):
                 return False
             return all(isinstance(q, MXFP8Quantizer) for q in input_quantizers) or all(
                 isinstance(q, NVFP4Quantizer) and q.with_rht for q in input_quantizers

--- a/transformer_engine/pytorch/ops/basic/grouped_linear.py
+++ b/transformer_engine/pytorch/ops/basic/grouped_linear.py
@@ -790,8 +790,14 @@ class GroupedLinear(BasicOperation):
             return False
         if with_quantized_compute:
             # FP8 per-tensor current scaling runs on the Hopper and Blackwell grouped GEMM
-            # path; the compute-capability range was already checked above.
+            # path; the compute-capability range was already checked above. On Hopper it
+            # requires cuBLAS 13.5+; fall back to the legacy flow on older cuBLAS.
             if all(isinstance(q, Float8CurrentScalingQuantizer) for q in input_quantizers):
+                if (
+                    get_device_compute_capability() < (10, 0)
+                    and tex.get_cublasLt_version() < 130500
+                ):
+                    return False
                 return True
             # MXFP8 and NVFP4 grouped quantization kernels require Blackwell.
             if not (10, 0) <= get_device_compute_capability() <= (11, 0):

